metamanager match pod messages for edged

yz271544 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kubeedge/kubeedge/blob/master/CONTRIBUTING.md
2. Ensure you have added or ran the appropriate tests for your PR

-->

**What type of PR is this?**
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind test
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind failing-test
-->


**What this PR does / why we need it**:
In order to ensure that the function of edge's watch is not affected, add a function `podMatchByNodeNameOrHostIP` for metamanager::process to fix the issue of #3489 that the cloud application is scheduled to the edge;

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #3489

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

fisherxu commented on edge/pkg/metamanager/process.go:
```diff
@@ -132,7 +132,14 @@ func (m *metaManager) processInsert(message model.Message) {
+		}
+	default:
+		sendToEdged(&message, false)
+	}
```
We can wrap a function to avoid duplication of code

yz271544 replied:
yes, wrapped the a func sendToEdgedByNodeNameOrHostIP(message model.Message, resType string, isSync bool)

fisherxu commented on edge/pkg/metamanager/process.go:
```diff
@@ -477,3 +491,41 @@ func (m *metaManager) runMetaManager() {
+			}
+		}
+	}
+	return
```
We have the resourceType in the message, can identify pod using it. Besides, just nodename is enough? :)

yz271544 commented:
At present, I really just judged `spec.nodeName` and `status.hostIP`. 

The reason is that I checked those pods specified by affinity through the log, and these two fields will be automatically added; 
whether it is enough to meet our matching conditions, we really need to continue Inquire about relevant documents and materials, or a large number of scene tests; 

In order to pass the integration test, I modified `helpers.go` and added a `nodeName` field. 

At present, this PR is issued, and I hope to get help from the community. 
**Please consider it carefully.**

yz271544 commented:
RFC ?

fisherxu commented on edge/pkg/metamanager/process.go:
```diff
@@ -175,7 +176,7 @@ func (m *metaManager) processUpdate(message model.Message) {
 		resp := message.NewRespByMessage(&message, OK)
 		sendToEdged(resp, message.IsSync())
 	case cloudmodules.EdgeControllerModuleName, cloudmodules.DynamicControllerModuleName:
-		sendToEdged(&message, message.IsSync())
```
Should not modify the `message.IsSync()`?

yz271544 replied:
yes, typo, corrected

fisherxu commented on edge/pkg/metamanager/process.go:
```diff
@@ -477,3 +478,46 @@ func (m *metaManager) runMetaManager() {
+
+func sendToEdgedByNodeNameOrHostIP(message model.Message, resType string, isSync bool) {
+	switch resType {
+	case "pod":
```
Use the const same as in edged.

yz271544 replied:
OK, modified

fisherxu commented on edge/pkg/metamanager/process.go:
```diff
@@ -477,3 +478,46 @@ func (m *metaManager) runMetaManager() {
+}
+
+func podMatchByNodeNameOrHostIP(message model.Message, nodeName string) (isMatched bool) {
+	content := message.GetContent()
```
We can json.Unmarshal  the content to pod directly.

yz271544 replied:
OK, modified
